TripalCultivate/TripalCultivate-Phenotypes
# G2.53 trait importer

**Issue #53 Phenotypic Trait Importer**

## Motivation


This PR will upgrade Analyzed Phenotypes Trait Importer to Drupal 9/10.

## What does this PR do?

1. An instance of Tripal Importer Plugin
2. A traits service that will handle creation of cvterm and relevant relationships.
3. Implement Tripal Cultivate Validator Plugins GENUS, DATA FILE, HEADERS and TRAIT IMPORTER VALUES.
4. Fixing some duplicated terms in the module config YAML and adding some new ones as needed.
5. Fixing a bug in the file template service that broke it in the testing environment: make directories if they do not yet exist + renderPlain since outside of the typical render flow.
6. Removing create() from services which specify dependencies in the service yaml instead.
7. Adds automated testing to confirm all parts added.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

#### Kernel: PluginValidatorTest::testTraitImportValuePluginValidator
1. All columns present, all columns have value and headers match the header array
2. Empty columns
3. Unexpected Type value column (Quantitative or Qualitative only)
8. Duplicate trait (Trait Name + Method Short Name + Unit)
9. Test skip flag to skip this test - set to upcoming validation step

#### Kernel: ValidTraitTest
1. Creation of terms for trait, method and unit plus all relationships required between trait and method, method and unit, and supplemental metadata about the unit.
10. Test that trait, method and unit are inserted into the correct Controlled Vocabulary (cv) and Database (db) as configured for the test Genus.
11. Addition of metadata about the unit.
12. Test the getters can retrieve all parts of a newly inserted trait.

#### Kernel: TraitImporterFormTest
1. Tests that the form generated for this importer has the genus + file upload elements
2. Test that the upload description has a template link and an ordered list describing columns
3. Test that we get a message if the module is not configured yet.
4. Check that when a valid file and genus are selected, the form can be submitted without error.
5. Check that if an invalid organism or no organism are set that there is a form state error.

#### Kernel: TraitImporterFormValidateTest
1. Test the form submit + build with various files that have an error in the file.
    - correct header but no data rows
    - incorrect header with correct data
    - correct header but a duplicate trait-method-unit combo
2. For all of the above, check that the validation_result form element is now present as expected in the form array and that the status and sometimes message are set as expected.

#### Kernel: TraitImporterRunTest
1. Test the run functionality of the Trait Importer class with a valid sample data file.
2. Confirm that no exceptions are thrown.

#### Trait: PhenotypeImporterTestTrait
Creates a test trait to make it easier to create files and configure the module in the testing environment.

### Coverage of main files added in this PR:
<img width="777" alt="Screenshot 2024-02-08 at 2 55 36 PM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/bf74784a-699c-45a4-ba50-c8057cbcf92c">
<img width="1420" alt="Screenshot 2024-02-08 at 2 55 56 PM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/41513244-61e9-4008-ae86-652de924b04c">
<img width="1421" alt="Screenshot 2024-02-08 at 2 56 06 PM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/49a1be84-484b-44ad-9d3c-6db1e88dd369">


### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Create test a genus `INSERT INTO testchado.organism (genus, species) VALUES ('Tripalus, 'databasica');`
7. Create cvs this genus `INSERT INTO testchado.cv (name) VALUES ('Tripalus Traits'), ('Tripalus Methods'), ('Tripalus Units'), ('Tripalus Crop Ontology');`
8. Create db for this genus `INSERT INTO testchado.db (name) VALUES ('Tripalus References');`
13. Configure the module and set a cv and a db for the test genus.
14. Go to the Tripal Cultivate Phenotypes Traits Importer
at `site/admin/tripal/loaders/trpcultivate-phenotypes-traits-importer`
15. Download a template file and alter the headers to test  things such as

- only header: fails file validation
- empty file: fails file validation
- header with missing column, extra column, wrong column: should fail header validation
- duplicated trait-method-unit combination
- one trait with multiple methods
- two traits sharing a method
- one method with multiple units
- multiple methods with the same unit
- something other than qualitative or quantitative in the last column
- existing trait with a different description
